runtime/src/main/scala/akka/grpc/internal/Marshallers.scala

```diff
@@ -18,19 +18,33 @@ abstract class BaseMarshaller[T](val protobufSerializer: ProtobufSerializer[T])
     extends io.grpc.MethodDescriptor.Marshaller[T]
     with WithProtobufSerializer[T] {
   override def parse(stream: InputStream): T = {
-    val baos = new ByteArrayOutputStream(math.max(64, stream.available()))
-    val buffer = new Array[Byte](32 * 1024)
+    val buffer =
+      new Array[Byte](stream match {
+        case k: KnownLength => math.max(0, k.available()) // No need to oversize this if we already know the size
+        case _              => 32 * 1024
+      })
 
     // Blocking calls underneath...
     // we can't avoid it for the moment because we are relying on the Netty's Channel API
-    var bytesRead = stream.read(buffer)
-    while (bytesRead >= 0) {
-      baos.write(buffer, 0, bytesRead)
-      bytesRead = stream.read(buffer)
+    val bytes = stream.read(buffer, 0, buffer.length) match {
```

Comment:
Is this overload better than just `stream.read(buffer)`?

Reply:
According to the javadoc they are semantically equivalent, but to someone who just glaces over the code might be tricked into thinking that it just reads as much as possible without blocking—so I opted to make it clear that I want buffer.length bytes or EOF.